Updated code to be compatible with dbus-java 4.x

hypfvieh opened the pull request:
This is an update to your extension for dbus-java.
Due to refactorings and splitting of dbus-java into core and transports (so support of Java 16+ native unix sockets is possible), some changes on the SPI was also required.

The changes include the usage of `SocketChannel` instead of `Socket` (because dbus-java now uses `java.nio` internally) and also the package name change for the `IMessageProvider` SPI (which was moved to a sub-package).

Code compiles and unit tests run on my test machine using latest SNAPSHOT of dbus-java 4.x

[Edit]
Forgot to mention: Java 11 is now required as minimum Java version (as dbus-java 4.x also requires it)

rm5248 commented:
So the changes to this code seem to be fine, but there may be a bug in dbus-java.  I'm getting this exception near the end of the `NativeStressTest`:

```
23:06:36.393 [DBusConnection] DEBUG c.r.d.nativefd.NativeMessageReader - Got the data
23:06:36.393 [DBusConnection] DEBUG c.r.d.nativefd.NativeMessageReader - => MethodCall(0,2039) { Path=>/org/freedesktop/dbus/test/RemoteObjectImpl, Interface=>com.rm5248.dbusjava.test.NativeStressTest.RemoteObject, Member=>doSomething, Destination=>:1.1609, Sender=>:1.1620, Signature=>s } { :1.1620 }
Exception in thread "DBusConnection" java.util.concurrent.RejectedExecutionException: Task org.freedesktop.dbus.connections.AbstractConnection$1@7a79c956 rejected from java.util.concurrent.ThreadPoolExecutor@5bbc92aa[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 113]
	at java.base/java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2055)
	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:825)
	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1355)
	at org.freedesktop.dbus.connections.AbstractConnection.sendMessage(AbstractConnection.java:390)
	at org.freedesktop.dbus.connections.AbstractConnection.handleMessage(AbstractConnection.java:769)
	at org.freedesktop.dbus.connections.AbstractConnection.handleMessage(AbstractConnection.java:723)
	at org.freedesktop.dbus.connections.IncomingMessageThread.run(IncomingMessageThread.java:43)
23:06:36.397 [DBusConnection] DEBUG c.r.d.n.NativeMessageReader.native - Received message.  Ret: 0 iovlen: 1 controllen: 0
23:06:36.397 [DBusConnection] ERROR c.r.d.n.NativeMessageReader.native - Unknown endianess!
23:06:36.398 [DBusConnection] DEBUG c.r.d.n.NativeMessageReader.native - Received message.  Ret: 0 iovlen: 1 controllen: 0
```

This does not happen with my current master branch using any of the 3.x releases of dbus-java.  The oddest thing about this is the line that says 'Unknown ednianess!', as that is coming from the [JNI code](https://github.com/rm5248/dbus-java-nativefd/blob/401ec7e12cb8d39721710e2d7dbe3745e7266939/src/main/jni/native-message-reader.c#L162).  The exception also seems to be swallowed by dbus-java, so the test is still passing when it _looks_ like it shouldn't be passing.

hypfvieh commented:
It seems that there are multiple problems.

First I can confirm, no 'strange' behavor when using dbus-java 3.x.
So my guess is, that the issues are introduced by using java.nio.

About the `RejectedExecutionException`, it was not caught and was thrown
in another thread (IncomingMessageThread), so it was just logged somewhere without any
additional actions.

I updated dbus-java 4.x to catch these exceptions as well. 

I also changed the order of 'closing' when disconnecting a transport (call `close()` on it).
Now I will first close MessageReader/MessageWriter and then close the underlying transport socket.
This makes more sense in my eyes than closing the socket and then closing Reader/Writer.

The mentioned issues with 'Invalid endianess' and also the swallowed exceptions are now gone on my machine.
I tested with JDK 11 and 17, in both cases the tests do no fail and there is also no indication of issues
in the logging output.

rm5248 commented:
awesome.  Looks all good to me; merged.
